Tarea clase 6

DarioAdise opened the pull request.

DarioAdise committed:
Primer tarea clase 6

Esta es la tarea de hacer cálculos con las edades de la familia.
Incluye el bonus de borrar elementos y volver a cargar.

DarioAdise committed:
Primer tarea de clase 6

Agrega tarea de cálculos sobre la edad de los integrantes del grupo familiar

anarod13 commented on tareas/clase-6/index.html:
```diff
@@ -0,0 +1,42 @@
+    </div>
+    </form>
+
+<form id="formulario_ingreso_edades" style="width: 80%; border-width: 1px; border-color: blue; border-style: solid;""></form>
```
Fijate que las últimas comillas están de más, igual que en la línea 14

anarod13 commented on tareas/clase-6/index.html:
```diff
@@ -0,0 +1,42 @@
+
+<form id="formulario_ingreso_edades" style="width: 80%; border-width: 1px; border-color: blue; border-style: solid;""></form>
+
+</form>
```
Puede ser que esta etiqueta de cierre no tenga apertura?

DarioAdise replied:
Hola Ana, gracias por tomarte el tiempo de corregir el código! Veo que cometí algunos typos. Voy a revisar mejor antes de enviar el código. Está bien que te vaya respondiendo por aquí? O sólo debo corregir y resubir? No sé bien cómo es el proceso de review a través de github...

anarod13 replied:
Hola Dario! De ambas formas está ok, lo bueno de hacerlo por acá  es que te vas acostumbrando a manejarte con los pr. Revisé los cambios que hiciste y está todo ok!

anarod13 commented on tareas/clase-6/index.html:
```diff
@@ -0,0 +1,42 @@
+    <em id="menor_edad"></em><br>
+    <em id="promedio_edad"></em><br>
+
+</form>
```
Ten en cuenta que la etiqueta `form` hace referencia a una sección del html que funciona como un formulario, tiene etiquetas (labels), campos para ingresar datos (inputs) y una acción (generalmente un botón para submitearlos), por lo que si bien html te permite usarlo, lo correcto sería agrupar todos los inputs y el botón o input de acción relacionado con estos bajo un solo form, cómo hiciste en [este primer form](https://github.com/DarioAdise/introduccion-a-js/pull/4/files#diff-199d1929f721b383d5959ebff948a439de9622821d400f67f84191483b409bdeR14-R20) ,  en el que pusiste el campo en que ingresa el número de integrantes y el input relacionado en un solo form.
En la docu puedes leer un poco más al respecto ([form-HTML](https://developer.mozilla.org/es/docs/Web/HTML/Element/form))

DarioAdise replied:
Entendido! Voy a usar un div para agrupar elementos que no sean estrictamente un formulario.

anarod13 commented on tareas/clase-6/index.html:
```diff
@@ -0,0 +1,42 @@
+        <input type="number" id="cantidad_integrantes" /> 
+        <input type="button" value="Mostrar formularios" id="mostrar_formularios" />
+    </div>
+    </form>
```
En este caso el uso de la etiqueta `form` está perfecto :+1:

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+		}
+		return edad_grande;
+	}
+	
```
Tené cuidado con la indentación! Instalate el prettier y el eslint para que te corrijan eso automáticamente, va a hacerte más facil leer el código

DarioAdise replied:
Ah no los conocía! Recién los instalé desde las extensiones de VS Code.

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+    suma_edades = suma_edades + edades[i];
+  }
+  return suma_edades / edades.length;
+}
```
Dentro de una función puedes llamar a otras funciones (es lo ideal) pero no es correcto declarar una función dentro de otra, pues va en contra de uno de los principios claves de las buenas prácticas de programación, que es que cada función debe tener una sola responsabilidad (Single responsability principle).
Acordate que declarar una función es definir que operaciones va a realizar:
`function sumarNúmeros(numero1,numero2){
   return numero1 + numero2
   }`
Y llamarla es hacer que esta se ejecute con los parámetros que le pase   
   `sumarNumeros(5,8);`
   
En este caso lo que podrías hacer sería definir las funciones de calcular el mayor, el menor y el promedio por fuera y pasarle cómo parámetro un array que representaría las edades

DarioAdise replied:
Si, ahora veo que estaba metiendo la pata al declarar funciones dentro de una función más grande, pero me quedaba comodísimo jaja. Las voy a declarar afuera y pasarle el array como argumento.

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+
+
+
+document.querySelector("#mayor_edad").innerText = `La mayor edad del grupo familiar es ${calcularMayorEdad()}`;
+document.querySelector("#menor_edad").innerText = `La menor edad del grupo familiar es ${calcularMenorEdad()}`;
+document.querySelector("#promedio_edad").innerText = `El promedio de edad del grupo familiar es ${calcularPromedioEdad()}`;
```
La forma en que las llamas aquí está perfecto, solo tenés que ajustar la forma en que están declaradas

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+  
+  if (cajas[i].value <1 || cajas[i].value > 120){
+    alert("Debe ingresar una edad entre 1 y 120");
+    senial = 1;
```
Aunque la variable `senial` cumple el propósito que estás buscando el nombre que le diste no da ninguna pista de el tipo de dato que estás almacenando y  el 0 y 1 no indican una lógica clara de porque les asignas ese valor. 

Lo que podrías hacer es usar este dato como una variable de SI o NO (un booleano) y llamarla algo como `estanTodasLasEdadesIngresadas= false`

DarioAdise replied:
Tienes razón, eso de senial 0 o 1 es un resabio de cuando usaba viejos lenguajes de programación. Voy a acostumbrarme a usar booleanos, y mejorar cómo nombro las variables.

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+let cajas = document.querySelector("[id=formulario_ingreso_edades]").querySelectorAll("input");
+let senial = 0;
+
+for (i=0; i<cajas.length;i++){
+  
+  if (cajas[i].value <1 || cajas[i].value > 120){
+    alert("Debe ingresar una edad entre 1 y 120");
+    senial = 1;
+		return false;
```
Otra cosa que te puede ayudar a subir el nivel de tu código es recurrir a los métodos de un array. En este caso `Array.every()` ejecutaría el mismo procedimiento que estás ejecutando con este for y te ahorra varias líneas de código [Aquí](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Array/every) puedes estudiarlos

DarioAdise replied:
Si, estoy atento a los métodos de array y siempre que trabajo con un array veo si puedo aplicar alguno. En los desafíos de la clase 4 (algoritmos) he usado todos a los que les encontré funcionalidad, como .push, .pop, .slice, .concat, .splice, .slice, .reverse, etc... Pero a veces no encuentro cuál o cómo aplicarlos y caigo en el tradicional loop For. Voy a investigar más!

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+  //vuelvo a activar el boton de mostrar formularios
+  document.getElementById("mostrar_formularios").disabled = false;
+
+  let ems = document.querySelectorAll("em");
```
Tené en cuenta que el nombre de las variables debe indicar que información estás almacenando, cosas como `ems`, `cajas` o `divisor` no dan ninguna información y dificultan entender el código, te implica leer todo el archivo para saber a que se están refiriendo. Recuerda además nombrarla en camelCase (`menorEdad` en lugar de `menor_edad``)

DarioAdise replied:
Entendido, voy a mejor eso!

anarod13 commented on tareas/clase-6/tarea-clase-6.js:
```diff
@@ -5,12 +5,164 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+    ems[i].innerText = "";
+  }
+  
+  document.getElementById('formulario_cantidad_integrantes').reset();
```
Por un tema de prolijidad procura usar la misma nomenclatura o método en tu código, por ej, en este caso usaste tres formas distintas para acceder a un elemento por su id:
`document.getElementById('formulario_cantidad_integrantes')`
`document.querySelector("#mayor_edad")`
`document.querySelector("[id=formulario_ingreso_edades]")`

Cualquiera de las 3 es correcta, pero mantener el uso de una sola te ayuda a tener un código más limpio

DarioAdise replied:
Bien, a veces lo hago para probar distintos métodos de cómo llamar al elemento HTML, y a veces por despistado. Voy a estandarizar eso!

DarioAdise committed:
Correcciones a la primer tarea de Clase 6

Corrijo varios errores indicados por Ana y agrego nuevas validaciones.

DarioAdise committed:
Tarea de cálculo de salarios

Agrega la segunda tarea de la clase que realiza cálculos sobre los salarios de un grupo familiar
